### PR TITLE
14.0 t0290

### PR DESCRIPTION
**Pull request by auclerc**

### CODE REVIEW

File: partner_compassion/models/partner_compassion.py [hunk 1]
```diff
@@ -522,6 +529,59 @@ def action_view_partner_invoices(self):
         action["context"] = {"search_default_partner_id": self.id}
         return action
 
+    def check_phone_and_mobile(self, vals):
+        all_phone_destination_codes = [
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]: Where does this list come from? Maybe a comment would be helpful here.

File: partner_compassion/models/partner_compassion.py [hunk 2]
```diff
@@ -522,6 +529,59 @@ def action_view_partner_invoices(self):
...
+            52,
+            55,
+            56,
+            58,
+            61,
+            62,
+            71,
+        ]
+
+        all_mobile_destination_codes = [74, 75, 76, 77, 78, 79]
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]: Where does it come from?

File: partner_compassion/models/partner_compassion.py [hunk 3]
```diff
@@ -522,6 +529,59 @@ def action_view_partner_invoices(self):
...
+
+        phone = vals.get("phone")
+        mobile = vals.get("mobile")
+
+        if phone:
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]: Maybe we should also check if the partner country is switzerland before doing the check. We may sometimes create partners from other countries that may have a different phone format. Or we can maybe validate the phone for any country with the phonenumbers library?

File: partner_compassion/models/partner_compassion.py [hunk 4]
```diff
@@ -522,6 +529,59 @@ def action_view_partner_invoices(self):
...
+            parsed_phone = phonenumbers.parse(phone, "CH")
+            if not phonenumbers.is_valid_number(parsed_phone):
+                raise UserError(_("Phone number is not valid."))
+            phone_national_destination_code = int(str(parsed_phone.national_number)[:2])
+            if phone_national_destination_code in all_mobile_destination_codes:
+                vals["mobile"] = phone
+                if not mobile:
+                    vals["phone"] = False
+
+        if mobile:
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]: Same remark about checking the country.

File: partner_compassion/models/partner_compassion.py [hunk 5]
```diff
@@ -522,6 +529,59 @@ def action_view_partner_invoices(self):
...
+        all_mobile_destination_codes = [74, 75, 76, 77, 78, 79]
+
+        phone = vals.get("phone")
+        mobile = vals.get("mobile")
+
+        if phone:
+            parsed_phone = phonenumbers.parse(phone, "CH")
+            if not phonenumbers.is_valid_number(parsed_phone):
+                raise UserError(_("Phone number is not valid."))
+            phone_national_destination_code = int(str(parsed_phone.national_number)[:2])
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]: I don't understand why you need such conversions. What is the format of `parsed_phone.national_number`?

File: partner_compassion/models/partner_compassion.py [hunk 6]
```diff
@@ -522,6 +529,59 @@ def action_view_partner_invoices(self):
...
+            if phone_national_destination_code in all_mobile_destination_codes:
+                vals["mobile"] = phone
+                if not mobile:
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]: Why is this `if` needed, shouldn't we remove the `phone` value in any case, if it's moved in the `mobile` field?

File: partner_compassion/models/partner_compassion.py [hunk 7]
```diff
@@ -522,6 +529,59 @@ def action_view_partner_invoices(self):
...
+            parsed_mobile = phonenumbers.parse(mobile, "CH")
+            if not phonenumbers.is_valid_number(parsed_mobile):
+                raise UserError(_("Mobile number is not valid."))
+            mobile_national_destination_code = int(
+                str(parsed_mobile.national_number)[:2]
+            )
+            if mobile_national_destination_code in all_phone_destination_codes:
+                vals["phone"] = mobile
+                if not phone:
+                    vals["mobile"] = False
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]: Maybe we should better check if a `phone` value is already set and not override it, and also check that we didn't move the `phone` field value and avoid deleting it afterward (you should introduce the bool var for that purpose).
```suggestion
                if not vals.get("phone"):
                    vals["phone"] = mobile
                    if not phone_moved_to_mobile:
                        vals["mobile"] = False
```

File: partner_compassion/models/partner_compassion.py [hunk 8]
```diff
@@ -522,6 +529,59 @@ def action_view_partner_invoices(self):
...
+
+        return vals
+
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]: The return is not needed, the dictionary is modified in place.

**Review by ecino — CHANGES_REQUESTED** [review]

File: partner_compassion/models/partner_compassion.py (hunk 5 above)
Comment (auclerc) [reply partner_compassion/models/partner_compassion.py]: To use the .is_valid_number(), we have to use a parsed number.

File: partner_compassion/models/partner_compassion.py (hunk 6 above)
Comment (auclerc) [reply partner_compassion/models/partner_compassion.py]: Yes, correct

File: partner_compassion/models/partner_compassion.py [hunk 9]
```diff
@@ -522,6 +529,66 @@ def action_view_partner_invoices(self):
...
+            61,
+            62,
+            71,
+        ]
+
+        # Destination codes are the first two digits of a mobile phone number
+        all_swiss_mobile_destination_codes = [74, 75, 76, 77, 78, 79]
+
+        # Check if the partner country is Switzerland
+        if vals.get('country_id') == 43 or self.country_id == 43:
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]:
```suggestion
        swiss_country = self.env.ref("base.ch")
        if vals.get('country_id') == swiss_country.id or self.mapped("country_id") == swiss_country:
```

File: partner_compassion/models/partner_compassion.py [hunk 10]
```diff
@@ -522,6 +529,66 @@ def action_view_partner_invoices(self):
...
+
+            if phone:
+                parsed_phone = phonenumbers.parse(phone, "CH")
+                if not phonenumbers.is_valid_number(parsed_phone):
+                    raise UserError(_("Phone number is not valid."))
+                phone_national_destination_code = int(str(parsed_phone.national_number)[:2])
+                if phone_national_destination_code in all_swiss_mobile_destination_codes:
+                    vals["mobile"] = phone
+                    phone_moved_to_mobile = True
+                    # if not mobile:
```
Comment (ecino) [inline partner_compassion/models/partner_compassion.py]:
```suggestion
```

**Review by ecino — CHANGES_REQUESTED** [review]

**Comment by ecino** [comment]
Thank you, I merged manually (for fixing conflicts).

**Pull request closed without merging**